ee/control/control.go

```diff
@@ -0,0 +1,142 @@
+package control
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// ControlService is the main object that manages the control service. It is responsible for fetching
+// and caching control data, and updating consumers and subscribers.
+type ControlService struct {
+	logger          log.Logger
+	cancel          context.CancelFunc
+	requestInterval time.Duration
+	consumers       map[string]consumer
+	subscribers     map[string][]subscriber
+	data            dataProvider
+	lastFetched     map[string]string
+}
+
+// consumer is an interface for something that consumes control server data updates.
+type consumer interface {
+	Update(io.Reader)
+}
+
+// subscriber is an interface for something that wants to be notified when a subsystem has been updated.
+type subscriber interface {
+	Ping()
+}
+
+// dataProvider is an interface for something that can retrieve control data. Authentication, HTTP,
+// file system access, etc. lives below this abstraction layer.
+type dataProvider interface {
+	Get(subsystem, cachedETag string) (etag string, data io.Reader, err error)
+}
+
+func NewControlService(logger log.Logger, data dataProvider, opts ...Option) *ControlService {
+	cs := &ControlService{
+		logger:          logger,
+		requestInterval: 60 * time.Second,
+		consumers:       make(map[string]consumer),
+		subscribers:     make(map[string][]subscriber),
+		data:            data,
+	}
+
+	for _, opt := range opts {
+		opt(cs)
+	}
+
+	return cs
+}
+
+func (cs *ControlService) Start(ctx context.Context) {
+	ctx, cs.cancel = context.WithCancel(ctx)
+	requestTicker := time.NewTicker(cs.requestInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-requestTicker.C:
+			cs.Fetch()
+		}
+	}
+}
+
+func (cs *ControlService) Stop() {
+	cs.cancel()
+}
+
+// controlResponse is the payload received from the control server
+type controlResponse struct {
+	// TODO: This is a temporary and simple data format for phase 1
+	Message string `json:"message,omitempty"`
+	Err     string `json:"error,omitempty"`
+}
+
+// Performs a retrieval of the latest control server data, and notifies observers of updates.
+func (cs *ControlService) Fetch() error {
+	_, data, err := cs.data.Get("", "")
```

Comment:
None of the etagging is plumbed here, correct?

Reply:
👍 removed etag stuff overall